src/machine.rs

```diff
@@ -11,6 +11,8 @@ use std::fmt;
 use log::trace;
 use rand::rngs::StdRng;
 use rand::SeedableRng;
+use std::collections::hash_map::Entry;
```

Comment:
Please keep `std` imports in a separate group.

Reply:
Also `Entry` is such a generic type name... maybe import just the `hash_map` module, and use `hash_map::Entry` below?